### PR TITLE
Decorate a class

### PR DESCRIPTION
**Pull request by zakiyousufi**
**Built With**
- Languages: Ruby
- Frameworks: N/A
- Technologies Used: GIT, GITHUB, LINTERS

_In this PR I have added the classes Person, Student, and Teacher according to these requirements_ 

- **Interface**
> - Create a class Nameable.
> > - Implement a method called `correct_name `that will raise a `NotImplementedError`.

- **Turn your Person class to Nameable**
> - Make sure that your Person class inherits from Nameable
> - Make sure that this class has a method` correct_name` implemented. It should simply return the name attribute.

- **Prepare base Decorator**
> - Make sure that it inherits from Nameable.
> - In the constructor assign a nameable object from params to an instance variable.
> - Implement the `correct_name` method that returns the result of the correct_name method of the @nameable.

- **Prepare CapitalizeDecorator and TrimmerDecorator**
> - For the CapitalizeDecorator:
> > - Create a class that inherits from the base Decorator class.
> > - Implement a method correct_name that capitalizes the output of @nameable.correct_name.
> - For the TrimmerDecorator:
> > - Create a class that inherits from the base Decorator class.
> > - Implement a method correct_name that makes sure that the output of @nameable.correct_name has a maximum of 10 characters. If it's longer it should trim the word.

### CODE REVIEW

File: classes/person.rb [hunk 1]
```diff
@@ -1,7 +1,10 @@
-class Person
+require_relative 'Nameable'
+
```
Comment (codecaiine) [inline classes/person.rb]: - [ ] From the project requirement your `Person` class needs to have a method `correct _name` . Please fix it :+1:

File: classes/decorate.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  def correct_name
+    @nameable.correct_name.capitalize
+  end
+end
+
+class TrimmerDecorator < BaseDecorator
+  def correct_name
+    @nameable.correct_name[0, 10] if @nameable.correct_name.length > 10
+  end
+end
```
Comment (codecaiine) [inline classes/decorate.rb:24]: - [ ] [OPTIONAL] Generally it is best practice to keep your classes in separate files that include the class name in the file name. This keeps your project better organized and makes it easier for someone else to navigate your structure. Please fix it :smile:

**Review by codecaiine — CHANGES_REQUESTED** [review]
Hi @zakiyousufi , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

## Highlights
- Good project implementation :heavy_check_mark: 
- No linters errors :heavy_check_mark: 
- Descriptive PR :heavy_check_mark: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me (@codecaiine) in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: classes/decorate.rb (hunk 2 above)
Comment (zakiyousufi) [reply classes/decorate.rb:24]: @codecaiine  
Thanks for your perfect review and wonderful feedback

File: classes/person.rb [hunk 3]
```diff
@@ -1,24 +1,29 @@
-class Person
-  attr_accessor :id, :name, :age, :parent_permission
+require_relative 'Nameable'
 
-  def initialize(age, name = 'unknown', parent_permission: true)
-    @id = Random.rand(1...1000)
+class Person < Nameable
```
Comment (BertrandConxy) [inline classes/person.rb:2]: - [ ] Good job on making changes, 👏 As suggested by the previous reviewer, It looks like you have not yet implemented the `correct_name` method. See this from the project requirements 👇 
image

HINT: you just need to define the `correct_name` method inside here and then simply return the name variable 😃 👍

**Review by BertrandConxy — CHANGES_REQUESTED** [review]
Hi @zakiyousufi, 

Great work on making the changes requested by a previous reviewer 👏 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights 💪 
- No linter errors
- correct flow is used
- good improvement based on the previous reviewer's suggestions
- good PR


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @BertrandConxy in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by cwaku — APPROVED** [review]
Hi @zakiyousufi 

Your project is complete! There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
 - [x] Correct GitHub Flow 👍🏾 
 - [x] Good commit history 🥇 
 - [x] Good PR title/description 💯 
 - [x] All linters passed 🥇 
 


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. The previous reviewer simply missed some of them, and addressing them will improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561), you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**